### DISCUSSION TITLE
FourPort Subnetwork to TwoPort Issues on skrf Version 0.30.0

### DISCUSSION THREAD

**Discussion by denzchoe** [scikit-rf/scikit-rf#994] [Q&A] [answered]
## Summary:
A Four Port file after subnetworking to 2 ports matches in S parameters, but doesn't match in Y parameters. Is this a bug?

## Description:
I am attaching a sample _.s4p_ file here.
I ran the bottom script.

```python
#%% Issue
import skrf as rf

fourportFile = rf.Network('fourportFile.s4p')
twoportFile = rf.subnetwork(fourportFile, [0,1])
twoportFile.name = 'twoportFile'

## Y Parameters
plt.figure()
fourportFile.plot_y_db(0,0)
fourportFile.plot_y_db(1,1)
fourportFile.plot_y_db(1,0)
twoportFile.plot_y_db(ls='--')
plt.title('Y Parameters Comparison')
plt.legend(fontsize = title_fontsize*0.25)


## S Parameters
plt.figure()
fourportFile.plot_s_db(0,0)
fourportFile.plot_s_db(1,1)
fourportFile.plot_s_db(1,0)
twoportFile.plot_s_db(ls='--')
plt.title('S Parameters Comparison')
plt.legend(fontsize = title_fontsize*0.25)


#%% Logical Test 1
fourportFile.y_db[:,1,1] == twoportFile.y_db[:,1,1]
#%% Logical Test 2
fourportFile.s_db[:,1,1] == twoportFile.s_db[:,1,1]


```

You will obtain the following plots and logic comparison;

![image](https://github.com/scikit-rf/scikit-rf/assets/26157143/cbf37044-2c27-4b74-92ba-b60816cb59e8)
![image](https://github.com/scikit-rf/scikit-rf/assets/26157143/90c539c6-9371-4d45-ac54-ab028eb54c49)
![image](https://github.com/scikit-rf/scikit-rf/assets/26157143/9c2e8be6-f164-4e09-bb79-795eb34958e3)
![image](https://github.com/scikit-rf/scikit-rf/assets/26157143/2246ca35-ee1e-41ff-afab-8c8ed4b2b132)

<br> 

Why does S Parameters comparison appears to show they are identical but Y Parameters & etc shows it does not after running `subnetwork`?

I have tested this on scikit-rf version `0.30.0` & `0.24.1`. Results are both similar.

<br>
<br>

## Current Status (2024-01-04):
When I tried to rebuild the new 2 port Network using Y-Parameters extracted from fourportFile, my Y Parameters are aligned graphically on the graph. The new S-Parameters (neWtwoportFile) now shifted however.

```python
# Before creating the scikit-rf Network object, one must forge the Frequency and S-matrix:
freq2 = rf.Frequency.from_f(twoportFile.f, unit='Hz')

NwkP = np.zeros((len(twoportFile.f), 2, 2), dtype=complex)
NwkP[:,0,0] = fourportFile.y[:,0,0]
NwkP[:,0,1] = fourportFile.y[:,0,1]
NwkP[:,1,0] = fourportFile.y[:,1,0]
NwkP[:,1,1] = fourportFile.y[:,1,1]


# constructing Network object
neWtwoportFile = rf.Network(frequency=freq2, y = NwkP, unit='GHz')
neWtwoportFile.name = 'neWtwoportFile'
neWtwoportFile.frequency.unit = 'GHz'


## Y Parameters
plt.figure()
fourportFile.plot_y_db(0,0, lw = 1.5)
fourportFile.plot_y_db(1,1, lw = 1.5)
fourportFile.plot_y_db(1,0, lw = 1.5)
twoportFile.plot_y_db(ls='--', lw = 2.5, markevery=15)
neWtwoportFile.plot_y_db(ls='--')
plt.title('Y Parameters Comparison')
plt.legend(fontsize = title_fontsize*0.25)


## S Parameters
plt.figure()
fourportFile.plot_s_db(0,0, lw = 1.5)
fourportFile.plot_s_db(1,1, lw = 1.5)
fourportFile.plot_s_db(1,0, lw = 1.5)
twoportFile.plot_s_db(ls='--', lw = 2.5, markevery=15)
neWtwoportFile.plot_s_db(ls='--')
plt.title('S Parameters Comparison')
plt.legend(fontsize = title_fontsize*0.25)

```
![image](https://github.com/scikit-rf/scikit-rf/assets/26157143/55749cea-48d0-47f6-8003-f1273eb57b54)
![image](https://github.com/scikit-rf/scikit-rf/assets/26157143/259bc24a-cdbd-4583-b3f5-dbb182b059cc)

 

---

## Attachment: 
Please convert the attachment's `.txt` to `.s4p` to reproduce the plots
- [fourportFile.txt](https://github.com/scikit-rf/scikit-rf/files/13821486/fourportFile.txt)

**Comment by denzchoe** [comment]
Main Thread Content (above) updated for clarity in context continuity. 
See `Current Status (2024-01-04)` Header. 
Click **view it on GitHub**

**Reply by jhillairet** [reply]
@denzchoe, how about latest versions?

**Reply by denzchoe** [reply]
ah... I have yet to try it. Overwhelmed with work, going to take a while to circle back to this topic for me.
I have yet to update my scikit-rf package in my conda env.

Did you manage to test it with the latest version?
My code and Touchstone file was attached as above

**Comment by Vinc0110** [comment]
Is this maybe related to the discussion in issue #977? After all, I think the network will always be created from S parameters and then converted to Y or Z. Could these rather small differences result from numeric errors in the conversion process?

**Comment by Asachoo** [comment]
Given that Y-parameters measured under the condition of **[short-circuiting](https://en.wikipedia.org/wiki/Admittance_parameters)** other ports, using `Network.subnetwork()` only crops the S-parameters, it implies that the eliminated ports are the **matching conditions**.

Thus, it seems reasonable that inconsistent Y-parameter results are obtained in your example @denzchoe .

I conducted a test by shorting ports 2 and 3, and the same Y-parameters were obtained. However, it's clear that the S-parameters of `twoportFile` can no longer be consistent with `fourportFile`.

```python3
import skrf as rf
from matplotlib import pyplot as plt

fourportFile = rf.Network('fourportFile.s4p')
freq = fourportFile.frequency
media = rf.media.DefinedGammaZ0(frequency=freq)
short_ntw = media.short(1)
twoportFile = rf.connect(fourportFile, 2, short_ntw, 0) # Short port 3
twoportFile = rf.connect(twoportFile, 2, short_ntw, 0) # Short port 4

twoportFile.name = "twoportFile"

## Y Parameters
plt.figure()
fourportFile.plot_y_db(0, 0)
fourportFile.plot_y_db(1, 1)
fourportFile.plot_y_db(1, 0)
twoportFile.plot_y_db(0,0, ls="--")
twoportFile.plot_y_db(1,1, ls="--")
twoportFile.plot_y_db(1, 0, ls="--")
plt.title("Y Parameters Comparison")

print(np.allclose(fourportFile.y_db[:,1,1], twoportFile.y_db[:,1,1]))

plt.show()

```
And here's the output and comparison plot:
```shell
True
```
![Figure_1](https://github.com/user-attachments/assets/7181dd88-350b-498d-a521-f6bd6fd8acae)

**Reply by Asachoo** [reply, marked as answer]
Has this analysis effectively resolved the discussion? @denzchoe
